eigenpro/models/sharded_kernel_machine.py

```diff
@@ -11,157 +11,159 @@
-
-    Returns:
-        torch.Tensor: tensor of shape [n_samples, n_outputs].
-    """
-
-    x_broadcast = self.device(x)
-    with ThreadPoolExecutor() as executor:
-      predictions = [executor.submit(self.shard_kms[i].forward, x_broadcast[i],train=train)
-                     for i in range(self.n_devices)]
-    results = [k.result() for k in predictions]
-
-    p_all = 0
-    k_centers_batch_all = []
-    for i,r in enumerate(results):
-      p_all =  p_all + r.to(self.device.device_base)
-      k_centers_batch_all.append(self.shard_kms[i].lru.get('k_centers_batch'))
-      self.shard_kms[i].lru.cache.clear()
-      torch.cuda.empty_cache()
-      del r
-
-    if train:
-      self.lru.put('k_centers_batch',k_centers_batch_all )
-
-    del x_broadcast,k_centers_batch_all,results,x
-    torch.cuda.empty_cache()
-
-
-    return p_all
-
-  def add_centers(self, centers: torch.Tensor, weights: Optional[torch.Tensor] = None
-                  ,nystrom_centers = False) -> None:
-    """Adds new centers and weights.
-
-    Args:
-      centers: Kernel centers of shape [n_centers, n_features].
-      weights: Weight parameters corresponding to the centers of shape
-               [n_centers, n_output].
-
-    Returns:
-      center_weights: Weight parameters corresponding to the added centers.
-    """
-    centers_gpus_list = self.device(centers, strategy="divide_to_gpu")
-    center_weights = weights if weights is not None else torch.zeros(
-        (centers.shape[0], self.n_outputs))
-    weights_gpus_list = self.device(center_weights, strategy="divide_to_gpu")
-
-
-    with ThreadPoolExecutor() as executor:
-      outputs = [executor.submit(self.shard_kms[i].add_centers, centers_gpus_list[i]
-                           , weights_gpus_list[i],nystrom_centers=nystrom_centers) for i in range(self.n_devices)]
-
-    for outputs in as_completed(outputs):
-      try:
-        # Retrieve the result of the future
-        _ = outputs.result()
-        # Process the result if necessary
-      except Exception as exc:
-        # Handle exceptions from within the thread
-        print(f"A thread caused an error: {exc}")
-        raise  # Reraising the exception will stop the program
-
-
-    return center_weights
-  def update_by_index(self, indices: torch.Tensor,
-                      delta: torch.Tensor,
-                      nystrom_update:bool = False) -> None:
-    """Update the model weights by index.
+    """Kernel machine that shards its computation across multiple devices."""
+
+    def __init__(self, kms: List[km.KernelMachine], device: device.Device):
+        self.device = device
+        self.shard_kms = kms
+        self.n_devices = len(kms)
+        self.n_machines = len(kms)
+        self.lru = cache.LRUCache()
+        super().__init__(
+            kms[0].kernel_fn,
+            kms[0].n_outputs,
+            sum(km.size for km in kms)
+        )
 
-    Here we assume that only the first block is trainable.
+    @property
```

Comment:
change the tab to 2 spaces here for consistency with the rest of the projects.
Also, using a different tabbing make it impossible to see what changes you have made.